### PR TITLE
feat(gitlab): add missing types to implement caching in GitLab pipelines

### PR DESCRIPTION
**Pull request by nbyl**
#1357 added the types to build GitLab CI pipelines. The [base schema](https://gitlab.com/gitlab-org/gitlab/-/raw/master/app/assets/javascripts/editor/schema/ci.json) unfortunately broke the implementation of caching into two entities. In the [manual](https://docs.gitlab.com/ee/ci/yaml/#cache), this is only one entity.

In the current state, caching between pipeline runs is actually not very usable. You can only express when to cache, not what. 

This PR adds the types needed to generate cache statements according to the GitLab reference for [GitLab 14.10](https://docs.gitlab.com/14.10/ee/ci/yaml/).

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/gitlab/configuration.test.ts [hunk 1]
```diff
@@ -127,3 +127,41 @@ test("respect the original format when adding global variables", () => {
...
+    default: {
+      cache: {
+        paths: ["node_modules"],
+        key: "${CI_COMMIT_REF_SLUG}",
+      },
+    },
+  });
+  const snapshot = synthSnapshot(p);
+  // THEN
+  expect(snapshot).toMatchSnapshot();
```
Comment (Chriscbr) [inline test/gitlab/configuration.test.ts:146]: Please scope these down to the appropriate changes
```suggestion
  expect(snapshot[".gitlab/ci-templates/foo.yml"]).toMatchSnapshot();
```

File: test/gitlab/configuration.test.ts [hunk 2]
```diff
@@ -127,3 +127,41 @@ test("respect the original format when adding global variables", () => {
...
+        key: {
+          files: ["Gemfile.lock", "package.json"],
+          prefix: "${CI_COMMIT_REF_SLUG}",
+        },
+      },
+    },
+  });
+  const snapshot = synthSnapshot(p);
+  // THEN
+  expect(snapshot).toMatchSnapshot();
```
Comment (Chriscbr) [inline test/gitlab/configuration.test.ts:166]:
```suggestion
  expect(snapshot[".gitlab/ci-templates/foo.yml"]).toMatchSnapshot();
```

**Review by Chriscbr — CHANGES_REQUESTED** [review]

File: test/gitlab/__snapshots__/configuration.test.ts.snap [hunk 3]
```diff
@@ -0,0 +1,5 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`adds correct entries for file-based caching 1`] = `undefined`;
+
+exports[`adds correct entries for path-based caching 1`] = `undefined`;
```
Comment (Chriscbr) [inline test/gitlab/__snapshots__/configuration.test.ts.snap:5]: @nbyl These should not be empty. I think we should be capturing a different file name?

Comment (nbyl) [reply test/gitlab/__snapshots__/configuration.test.ts.snap]: @Chriscbr Sorry, i missed that. I've updated the file path and snapshots.

**Review by Chriscbr — APPROVED** [review]
👍

**Pull request merged**